adal/src/main/java/com/microsoft/aad/adal/ADALError.java

```diff
@@ -638,6 +634,38 @@ public String getLocalizedDescription(Context context) {
             return resources.getString(resources.getIdentifier(this.name(), "string",
                     context.getPackageName()));
         }
-        return mCommonADALError.getDescription();
+        return mDescription;
+    }
+    private static final Map<String, ADALError> exceptionMap = new HashMap<String,ADALError>();
+
+    static {
+        exceptionMap.put(CommonCoreException.ANDROIDKEYSTORE_FAILED, ADALError.ANDROIDKEYSTORE_FAILED);
+        exceptionMap.put(CommonCoreException.AUTHORITY_URL_NOT_VALID, ADALError.DEVELOPER_AUTHORITY_IS_NOT_VALID_INSTANCE);
+        exceptionMap.put(CommonCoreException.AUTHORITY_VALIDATION_NOT_SUPPORTED, ADALError.DEVELOPER_AUTHORITY_CAN_NOT_BE_VALIDED);
+        exceptionMap.put(CommonCoreException.DECRYPTION_ERROR, ADALError.DECRYPTION_FAILED);
+        exceptionMap.put(CommonCoreException.DEVICE_NETWORK_NOT_AVAILABLE, ADALError.DEVICE_CONNECTION_IS_NOT_AVAILABLE);
+        exceptionMap.put(CommonCoreException.ENCRYPTION_ERROR, ADALError.ENCRYPTION_ERROR);
+        exceptionMap.put(CommonCoreException.INVALID_JWT, ADALError.JSON_PARSE_ERROR);
+        exceptionMap.put(CommonCoreException.IO_ERROR, ADALError.IO_EXCEPTION);
+        exceptionMap.put(CommonCoreException.JSON_PARSE_FAILURE, ADALError.JSON_PARSE_ERROR);
+        // TODO: :O
+        exceptionMap.put(CommonCoreException.MALFORMED_URL, ADALError.DEVELOPER_AUTHORITY_IS_NOT_VALID_URL);
+        exceptionMap.put(CommonCoreException.MULTIPLE_MATCHING_TOKENS_DETECTED, ADALError.AUTH_FAILED_USER_MISMATCH);
+        exceptionMap.put(CommonCoreException.USER_MISMATCH, ADALError.AUTH_FAILED_USER_MISMATCH);
+        exceptionMap.put(CommonCoreException.NO_SUCH_ALGORITHM, ADALError.DEVICE_NO_SUCH_ALGORITHM);
+        exceptionMap.put(CommonCoreException.UNSUPPORTED_ENCODING, ADALError.ENCODING_IS_NOT_SUPPORTED);
+
+        // The following mapping need a lot more brushing up
+        exceptionMap.put(CommonCoreServiceException.ACCESS_DENIED, ADALError.AUTH_REFRESH_FAILED_PROMPT_NOT_ALLOWED);
+        exceptionMap.put(CommonCoreServiceException.INVALID_INSTANCE, ADALError.AUTH_REFRESH_FAILED_PROMPT_NOT_ALLOWED);
+        exceptionMap.put(CommonCoreServiceException.INVALID_REQUEST, ADALError.SERVER_INVALID_REQUEST);
+        exceptionMap.put(CommonCoreServiceException.REQUEST_TIMEOUT, ADALError.SERVER_ERROR);
+        exceptionMap.put(CommonCoreServiceException.SERVICE_NOT_AVAILABLE, ADALError.SERVER_ERROR);
+        exceptionMap.put(CommonCoreServiceException.UNKNOWN_ERROR, ADALError.SERVER_ERROR);
+    }
+
+    public static ADALError fromCommon(final CommonCoreBaseException exception) {
+        // What do we do if errorCode is null?
```

Comment:
Possibly we create a new generic 'stuff went wrong but didn't report an error code' kind of exception? Something to handle the default case if the `exceptionMap` returns `null` on `get()` -- seems like the most sane thing I can think of, short of throwing a `RuntimeException` to indicate it's a programming error

Reply:
The more I think about it, the more I'm convinced we need to have _something_ happen when `null` is returned from `get()` -- it would be too easy to add a new `CommonCoreServiceException` and fail to update this mapping.

What if we [provided a setter on this enum](https://stackoverflow.com/questions/15302188/are-enums-allowed-to-have-setters-in-java) to track the `String` `errorCode` returned from `common` -- would we still have to maintain this mapping?